w3af/plugins/grep/csp.py

```diff
@@ -19,158 +19,163 @@
-                    occurrences = self._find_occurrences(csp_vuln.desc)
-
-                    if len(occurrences) > 1:
-                        # Shared vuln case
-                        v = Vuln(self.VULN_NAME, csp_vuln.desc,
-                                 csp_vuln.severity, occurrences,
-                                 self.get_name())
-                        v.set_url(vuln_store_item.url.base_url())
-
-                        vuln_already_reported.append(csp_vuln.desc)
-                    else:
-                        # Isolated vuln case
-                        v = Vuln(self.VULN_NAME, csp_vuln.desc,
-                                 csp_vuln.severity, vuln_store_item.resp_id,
-                                 self.get_name())
-                        v.set_url(vuln_store_item.url)
-
-                    # Report vuln
-                    self.kb_append(self, 'csp', v)
-                
-        # Cleanup
-        self._urls.cleanup()
-        self._vulns.cleanup()
 
-    def _find_occurrences(self, vuln_desc):
-        """
-        Internal utility function to find all occurrences of a vuln
-        into the global collection of vulns found by the plugin.
+        for csp in self.csps:
+            tmp = "URL: %s" % csp.response_url
+            for policy in csp.policies:
+                tmp += self.policy_tpl % (policy.pretty(), 
+                        "\n".join([v.desc for v in policy.find_vulns()]))
+            csp_vulns += tmp
         
-        @param vuln_desc: Vulnerability description.
-        @return: List of response ID for which the vuln is found.
+        if csp_vulns:
+            result += self.vuln_csp_tpl + csp_vulns
+        if self._responses_without_csp:
+            result += self.vuln_without_tpl % "\n".join(
+                    [r[1] for r in self._responses_without_csp])
+        result += self.vuln_footer_tpl
+
+        v = Vuln(self.VULN_NAME, result, severity.MEDIUM, 
+                [csp.response_id for csp in self.csps],
+                self.get_name())
+        self.kb_append(self, 'csp', v)
+
+    def get_options(self):
         """
-        list_resp_id = []
-
-        # Check input for quick exit
-        if vuln_desc is None or vuln_desc.strip() == "":
-            return list_resp_id
-       
-        # Parse vulns collection
-        ref = vuln_desc.lower().strip()        
-        for vuln_store_item in self._vulns:
-            for csp_directive_name, csp_vulns_list in vuln_store_item.csp_vulns.iteritems():
-                for csp_vuln in csp_vulns_list:        
-                    if csp_vuln.desc.strip().lower() == ref:
-                        if vuln_store_item.resp_id not in list_resp_id:
-                            list_resp_id.append(vuln_store_item.resp_id)
+        :return: A list of option objects for this plugin.
+        """
+        ol = OptionList()
 
-        return list_resp_id
+        d = 'Maximum number of sources in source list'
```

Comment:
This description doesn't make much sense if you don't read the code.

 * Change it to something meaningful
 * Add a help